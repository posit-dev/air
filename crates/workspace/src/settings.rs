//
// settings.rs
//
// Copyright (C) 2025 Posit Software, PBC. All rights reserved.
//
//

mod line_ending;

pub(crate) use line_ending::LineEnding;

use air_r_formatter::context::RFormatOptions;
use settings::IndentStyle;
use settings::IndentWidth;
use settings::LineWidth;
use settings::MagicLineBreak;

/// Resolved configuration settings used within air
///
/// May still require a source document to finalize some options, such as
/// `LineEnding::Auto` in the formatter.
#[derive(Clone, Debug, PartialEq, Eq, Default, serde::Serialize, serde::Deserialize)]
pub struct Settings {
    /// Settings to configure code formatting.
    pub format: FormatSettings,
}

#[derive(Clone, Debug, PartialEq, Eq, Default, serde::Serialize, serde::Deserialize)]
pub struct FormatSettings {
    pub indent_style: IndentStyle,
    pub indent_width: IndentWidth,
    pub line_ending: LineEnding,
    pub line_width: LineWidth,
    pub magic_line_break: MagicLineBreak,
}

impl FormatSettings {
    // Finalize `RFormatOptions` in preparation for a formatting operation on `source`
    pub fn to_format_options(&self, source: &str) -> RFormatOptions {
        RFormatOptions::new()
            .with_indent_style(self.indent_style)
            .with_indent_width(self.indent_width)
            .with_line_ending(self.line_ending.finalize(source))
            .with_line_width(self.line_width)
            .with_magic_line_break(self.magic_line_break)
    }
}
