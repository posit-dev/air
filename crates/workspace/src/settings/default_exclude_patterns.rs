use std::ops::Deref;
use std::sync::LazyLock;

use crate::file_patterns::DefaultFilePatterns;

/// The set of default exclude patterns
///
/// Importantly, default patterns apply with or without a physical `air.toml`, meaning
/// that we absolutely cannot use globs that only match absolute paths underneath a
/// `root`. We supply [ignore::gitignore::GitignoreBuilder] a `root` of the empty string
/// when builing [DEFAULT_EXCLUDE_PATTERNS], which means that nothing is stripped from
/// paths by `ignore` before performing a match (which is why the patterns can't match
/// absolute paths).
///
/// In a default pattern, you cannot use:
/// - Preceding `/`, like `/renv`, as that only matches `{root}/renv`
/// - Middle `/`, like `renv/*.R`, as that only matches `{root}/renv/*.R`
///
/// While not strictly necessary, to easily enforce this in tests all default patterns
/// must start with `**/`. Note that [ignore::gitignore::GitignoreBuilder] ensures that
/// `.git/` is equivalent to `**/.git/` and `cpp11.R` is equivalent to `**/cpp11.R`, so
/// this prefixing happens eventually anyways.
static DEFAULT_EXCLUDE_PATTERN_NAMES: &[&str] = &[
    // Directories
    // The trailing `/` prevents matching a non-directory file named, for example, `renv`.
    "**/.git/",
    "**/renv/",
    "**/revdep/",
    // Files
    "**/cpp11.R",
    "**/RcppExports.R",
    "**/extendr-wrappers.R",
    "**/import-standalone-*.R",
];

static DEFAULT_EXCLUDE_PATTERNS: LazyLock<DefaultFilePatterns> = LazyLock::new(|| {
    DefaultFilePatterns::try_from_iter(DEFAULT_EXCLUDE_PATTERN_NAMES.iter().copied())
        .expect("Can create default exclude patterns")
});

/// Typed wrapper around [DEFAULT_EXCLUDE_PATTERNS]
///
/// Allows for free creation of [DefaultExcludePatterns] structs without needing to clone
/// the global [DEFAULT_EXCLUDE_PATTERNS] object.
#[derive(Debug)]
pub struct DefaultExcludePatterns(&'static DefaultFilePatterns);

impl Default for DefaultExcludePatterns {
    /// Default exclude patterns
    ///
    /// Used in the [Default] method of [crate::settings::FormatSettings] to ensure that
    /// virtual `air.toml`s use the default exclude patterns.
    fn default() -> Self {
        Self(&DEFAULT_EXCLUDE_PATTERNS)
    }
}

impl Deref for DefaultExcludePatterns {
    type Target = DefaultFilePatterns;

    fn deref(&self) -> &Self::Target {
        self.0
    }
}

#[cfg(test)]
mod test {
    use crate::settings::default_exclude_patterns::DEFAULT_EXCLUDE_PATTERN_NAMES;
    use crate::settings::default_exclude_patterns::DefaultExcludePatterns;

    #[test]
    fn test_doublestar_default_patterns() {
        let _ = DEFAULT_EXCLUDE_PATTERN_NAMES
            .iter()
            .map(|pattern| assert!(pattern.starts_with("**/")));
    }

    #[test]
    fn test_default_exclude() -> anyhow::Result<()> {
        let default_patterns = DefaultExcludePatterns::default();

        assert!(default_patterns.matched("renv", true).is_some());
        assert!(default_patterns.matched("renv", false).is_none());
        assert!(
            default_patterns
                .matched_path_or_any_parents("renv/activate.R", false)
                .is_some()
        );

        assert!(default_patterns.matched("cpp11.R", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.R", false).is_some());

        assert!(
            default_patterns
                .matched("import-standalone-types-check.R", false)
                .is_some()
        );
        assert!(
            default_patterns
                .matched("R/import-standalone-foo.R", false)
                .is_some()
        );
        assert!(
            default_patterns
                .matched("pkg/R/import-standalone-foo.R", false)
                .is_some()
        );

        Ok(())
    }
}
