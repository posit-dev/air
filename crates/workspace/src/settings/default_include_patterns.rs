use std::ops::Deref;
use std::sync::LazyLock;

use crate::file_patterns::DefaultFilePatterns;

/// The set of default include patterns
///
/// See `DEFAULT_EXCLUDE_PATTERN_NAMES` for details on the exact structure of what can
/// be supplied here.
static DEFAULT_INCLUDE_PATTERN_NAMES: &[&str] = &[
    // R files with any filename at any depth
    "**/*.[R,r]",
];

static DEFAULT_INCLUDE_PATTERNS: LazyLock<DefaultFilePatterns> = LazyLock::new(|| {
    DefaultFilePatterns::try_from_iter(DEFAULT_INCLUDE_PATTERN_NAMES.iter().copied())
        .expect("Can create default include patterns")
});

/// Typed wrapper around [DEFAULT_INCLUDE_PATTERNS]
///
/// Allows for free creation of [DefaultIncludePatterns] structs without needing to clone
/// the global [DEFAULT_INCLUDE_PATTERNS] object.
#[derive(Debug)]
pub struct DefaultIncludePatterns(&'static DefaultFilePatterns);

impl Default for DefaultIncludePatterns {
    /// Default include patterns
    ///
    /// Used in the [Default] method of [crate::settings::FormatSettings] to ensure that
    /// virtual `air.toml`s use the default include patterns.
    fn default() -> Self {
        Self(&DEFAULT_INCLUDE_PATTERNS)
    }
}

impl Deref for DefaultIncludePatterns {
    type Target = DefaultFilePatterns;

    fn deref(&self) -> &Self::Target {
        self.0
    }
}

#[cfg(test)]
mod test {
    use crate::settings::default_include_patterns::DEFAULT_INCLUDE_PATTERN_NAMES;
    use crate::settings::default_include_patterns::DefaultIncludePatterns;

    #[test]
    fn test_doublestar_default_patterns() {
        let _ = DEFAULT_INCLUDE_PATTERN_NAMES
            .iter()
            .map(|pattern| assert!(pattern.starts_with("**/")));
    }

    #[test]
    fn test_default_include() -> anyhow::Result<()> {
        let default_patterns = DefaultIncludePatterns::default();

        assert!(default_patterns.matched("cpp11.R", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.R", false).is_some());

        assert!(default_patterns.matched("cpp11.r", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.r", false).is_some());

        assert!(default_patterns.matched("cpp11.py", false).is_none());
        assert!(default_patterns.matched("foo/cpp11.py", false).is_none());

        Ok(())
    }
}
