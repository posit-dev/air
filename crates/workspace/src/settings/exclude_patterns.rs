use std::ops::Deref;
use std::ops::DerefMut;
use std::path::Path;

use crate::file_patterns::FilePatterns;

#[derive(Debug, Clone)]
pub struct ExcludePatterns(FilePatterns);

impl ExcludePatterns {
    pub(crate) fn try_from_iter<'str, P, I>(root: P, patterns: I) -> anyhow::Result<Self>
    where
        P: AsRef<Path>,
        I: IntoIterator<Item = &'str str>,
    {
        Ok(Self(FilePatterns::try_from_iter(root, patterns)?))
    }
}

impl Deref for ExcludePatterns {
    type Target = FilePatterns;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for ExcludePatterns {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
