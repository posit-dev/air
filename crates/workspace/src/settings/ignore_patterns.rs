use std::ops::Deref;
use std::ops::DerefMut;
use std::path::Path;
use std::path::PathBuf;
use std::sync::LazyLock;

use crate::file_patterns::FilePatterns;

/// The set of default ignore patterns
///
/// Importantly, default patterns apply with or without a physical `air.toml`, meaning
/// that we absolutely cannot use globs that only match absolute paths underneath a
/// `root`. We supply [ignore::gitignore::GitignoreBuilder] a `root` of the empty string
/// when builing [DEFAULT_IGNORE_PATTERNS], which means that nothing is stripped from
/// paths by `ignore` before performing a match (which is why the patterns can't match
/// absolute paths).
///
/// In a default pattern, you cannot use:
/// - Preceding `/`, like `/renv`, as that only matches `{root}/renv`
/// - Middle `/`, like `renv/*.R`, as that only matches `{root}/renv/*.R`
///
/// While not strictly necessary, to easily enforce this in tests all default patterns
/// must start with `**/`. Note that [ignore::gitignore::GitignoreBuilder] ensures that
/// `.git/` is equivalent to `**/.git/` and `cpp11.R` is equivalent to `**/cpp11.R`, so
/// this prefixing happens eventually anyways.
static DEFAULT_IGNORE_PATTERN_NAMES: &[&str] = &[
    // Directories
    // The trailing `/` prevents matching a non-directory file named, for example, `renv`.
    "**/.git/",
    "**/renv/",
    "**/revdep/",
    // Files
    "**/cpp11.R",
    "**/RcppExports.R",
    "**/extendr-wrappers.R",
    "**/import-standalone-*.R",
];

static DEFAULT_IGNORE_PATTERNS: LazyLock<IgnorePatterns> = LazyLock::new(|| {
    IgnorePatterns::try_from_iter(PathBuf::new(), vec![], true)
        .expect("Can create default ignore patterns")
});

#[derive(Debug, Clone)]
pub struct IgnorePatterns(FilePatterns);

impl IgnorePatterns {
    pub(crate) fn try_from_iter<'str, P, I>(
        root: P,
        patterns: I,
        defaults: bool,
    ) -> anyhow::Result<Self>
    where
        P: AsRef<Path>,
        I: IntoIterator<Item = &'str str>,
    {
        if defaults {
            // Defaults come first, so user supplied patterns end up taking precedence
            let default_patterns = DEFAULT_IGNORE_PATTERN_NAMES.iter().copied();
            let patterns = default_patterns.chain(patterns);
            Ok(Self(FilePatterns::try_from_iter(root, patterns)?))
        } else {
            Ok(Self(FilePatterns::try_from_iter(root, patterns)?))
        }
    }
}

impl Default for IgnorePatterns {
    /// Default ignore patterns
    ///
    /// Used in the [Default] method of [crate::settings::FormatSettings] to ensure that
    /// virtual `air.toml`s use the default ignore patterns.
    fn default() -> Self {
        DEFAULT_IGNORE_PATTERNS.clone()
    }
}

impl Deref for IgnorePatterns {
    type Target = FilePatterns;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for IgnorePatterns {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
mod test {
    use crate::settings::ignore_patterns::IgnorePatterns;
    use crate::settings::ignore_patterns::DEFAULT_IGNORE_PATTERN_NAMES;

    #[test]
    fn test_doublestar_default_patterns() {
        let _ = DEFAULT_IGNORE_PATTERN_NAMES
            .iter()
            .map(|pattern| assert!(pattern.starts_with("**/")));
    }

    #[test]
    fn test_default_ignore() -> anyhow::Result<()> {
        let default_patterns = IgnorePatterns::default();

        assert!(default_patterns.matched("renv", true).is_some());
        assert!(default_patterns.matched("renv", false).is_none());
        assert!(default_patterns
            .matched_path_or_any_parents("renv/activate.R", false)
            .is_some());

        assert!(default_patterns.matched("cpp11.R", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.R", false).is_some());

        assert!(default_patterns
            .matched("import-standalone-types-check.R", false)
            .is_some());
        assert!(default_patterns
            .matched("R/import-standalone-foo.R", false)
            .is_some());
        assert!(default_patterns
            .matched("pkg/R/import-standalone-foo.R", false)
            .is_some());

        Ok(())
    }
}
