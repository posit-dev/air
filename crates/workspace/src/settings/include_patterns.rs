use std::ops::Deref;
use std::ops::DerefMut;
use std::path::Path;
use std::path::PathBuf;
use std::sync::LazyLock;

use crate::file_patterns::FilePatterns;

/// The set of default include patterns
///
/// See `DEFAULT_EXCLUDE_PATTERN_NAMES` for details on the exact structure of what can
/// be supplied here.
static DEFAULT_INCLUDE_PATTERN_NAMES: &[&str] = &[
    // R files with any filename at any depth
    "**/*.[R,r]",
];

static DEFAULT_INCLUDE_PATTERNS: LazyLock<IncludePatterns> = LazyLock::new(|| {
    IncludePatterns::try_from_iter(PathBuf::new(), vec![], true)
        .expect("Can create default include patterns")
});

#[derive(Debug, Clone)]
pub struct IncludePatterns(FilePatterns);

impl IncludePatterns {
    fn try_from_iter<'str, P, I>(root: P, patterns: I, defaults: bool) -> anyhow::Result<Self>
    where
        P: AsRef<Path>,
        I: IntoIterator<Item = &'str str>,
    {
        if defaults {
            // Defaults come first, so user supplied patterns end up taking precedence
            let default_patterns = DEFAULT_INCLUDE_PATTERN_NAMES.iter().copied();
            let patterns = default_patterns.chain(patterns);
            Ok(Self(FilePatterns::try_from_iter(root, patterns)?))
        } else {
            Ok(Self(FilePatterns::try_from_iter(root, patterns)?))
        }
    }
}

impl Default for IncludePatterns {
    /// Default include patterns
    ///
    /// Used in the [Default] method of [crate::settings::FormatSettings] to ensure that
    /// virtual `air.toml`s use the default include patterns.
    fn default() -> Self {
        DEFAULT_INCLUDE_PATTERNS.clone()
    }
}

impl Deref for IncludePatterns {
    type Target = FilePatterns;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for IncludePatterns {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
mod test {
    use crate::settings::include_patterns::IncludePatterns;
    use crate::settings::include_patterns::DEFAULT_INCLUDE_PATTERN_NAMES;

    #[test]
    fn test_doublestar_default_patterns() {
        let _ = DEFAULT_INCLUDE_PATTERN_NAMES
            .iter()
            .map(|pattern| assert!(pattern.starts_with("**/")));
    }

    #[test]
    fn test_default_include() -> anyhow::Result<()> {
        let default_patterns = IncludePatterns::default();

        assert!(default_patterns.matched("cpp11.R", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.R", false).is_some());

        assert!(default_patterns.matched("cpp11.r", false).is_some());
        assert!(default_patterns.matched("foo/cpp11.r", false).is_some());

        assert!(default_patterns.matched("cpp11.py", false).is_none());
        assert!(default_patterns.matched("foo/cpp11.py", false).is_none());

        Ok(())
    }
}
