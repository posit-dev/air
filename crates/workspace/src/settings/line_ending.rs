//
// line_ending.rs
//
// Copyright (C) 2025 Posit Software, PBC. All rights reserved.
//
//

use std::fmt;

#[derive(Copy, Clone, Debug, Eq, PartialEq, Default, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
pub enum LineEnding {
    /// The newline style is detected automatically on a file per file basis. Files with
    /// mixed line endings will be converted to the first detected line ending. Defaults
    /// to `\n` for a files that contain no line endings.
    #[default]
    Auto,

    /// Line endings will be converted to `\n` as is common on Unix.
    Lf,

    /// Line endings will be converted to `\r\n` as is common on Windows.
    Crlf,

    /// Line endings will be converted to `\n` on Unix and `\r\n` on Windows.
    Native,
}

impl fmt::Display for LineEnding {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Auto => write!(f, "Auto"),
            Self::Lf => write!(f, "LF"),
            Self::Crlf => write!(f, "CRLF"),
            Self::Native => write!(f, "Native"),
        }
    }
}

impl LineEnding {
    pub(crate) fn finalize(&self, source: &str) -> settings::LineEnding {
        match self {
            LineEnding::Lf => settings::LineEnding::Lf,
            LineEnding::Crlf => settings::LineEnding::Crlf,
            #[cfg(target_os = "windows")]
            LineEnding::Native => settings::LineEnding::Crlf,
            #[cfg(not(target_os = "windows"))]
            LineEnding::Native => settings::LineEnding::Lf,
            LineEnding::Auto => line_ending::infer(source),
        }
    }
}
