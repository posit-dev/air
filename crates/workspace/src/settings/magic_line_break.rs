//
// magic_line_break.rs
//
// Copyright (C) 2025 Posit Software, PBC. All rights reserved.
//
//

use std::fmt::Display;
use std::str::FromStr;

#[derive(Debug, Default, Clone, Copy, Eq, Hash, PartialEq)]
pub enum MagicLineBreak {
    /// Respect
    #[default]
    Respect,
    /// Ignore
    Ignore,
}

impl MagicLineBreak {
    /// Returns `true` if magic line breaks should be respected.
    pub const fn is_respect(&self) -> bool {
        matches!(self, MagicLineBreak::Respect)
    }

    /// Returns `true` if magic line breaks should be ignored.
    pub const fn is_ignore(&self) -> bool {
        matches!(self, MagicLineBreak::Ignore)
    }
}

impl FromStr for MagicLineBreak {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "respect" => Ok(Self::Respect),
            "ignore" => Ok(Self::Ignore),
            _ => Err("Unsupported value for this option"),
        }
    }
}

impl Display for MagicLineBreak {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MagicLineBreak::Respect => std::write!(f, "Respect"),
            MagicLineBreak::Ignore => std::write!(f, "Ignore"),
        }
    }
}

impl From<MagicLineBreak> for air_r_formatter::options::MagicLineBreak {
    fn from(value: MagicLineBreak) -> Self {
        match value {
            MagicLineBreak::Respect => air_r_formatter::options::MagicLineBreak::Respect,
            MagicLineBreak::Ignore => air_r_formatter::options::MagicLineBreak::Ignore,
        }
    }
}
