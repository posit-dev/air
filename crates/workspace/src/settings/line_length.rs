use std::fmt;
use std::num::NonZeroU16;

/// Validated value for the `line-length` formatter options
///
/// The allowed range of values is 1..=320
#[derive(Clone, Copy, Eq, PartialEq)]
pub struct LineLength(NonZeroU16);

impl LineLength {
    /// Maximum allowed value for a valid [LineLength]
    const MAX: u16 = 320;

    /// Return the numeric value for this [LineLength]
    pub fn value(&self) -> u16 {
        self.0.get()
    }
}

impl Default for LineLength {
    fn default() -> Self {
        Self(NonZeroU16::new(80).unwrap())
    }
}

impl std::fmt::Debug for LineLength {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        std::fmt::Debug::fmt(&self.0, f)
    }
}

impl fmt::Display for LineLength {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl<'de> serde::Deserialize<'de> for LineLength {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let value: u16 = serde::Deserialize::deserialize(deserializer)?;
        let line_length = LineLength::try_from(value).map_err(serde::de::Error::custom)?;
        Ok(line_length)
    }
}

/// Error type returned when converting a u16 or NonZeroU16 to a [`LineLength`] fails
#[derive(Clone, Copy, Debug)]
pub struct LineLengthFromIntError(u16);

impl std::error::Error for LineLengthFromIntError {}

impl TryFrom<u16> for LineLength {
    type Error = LineLengthFromIntError;

    fn try_from(value: u16) -> Result<Self, Self::Error> {
        match NonZeroU16::try_from(value) {
            Ok(value) => LineLength::try_from(value),
            Err(_) => Err(LineLengthFromIntError(value)),
        }
    }
}

impl TryFrom<NonZeroU16> for LineLength {
    type Error = LineLengthFromIntError;

    fn try_from(value: NonZeroU16) -> Result<Self, Self::Error> {
        if value.get() <= Self::MAX {
            Ok(LineLength(value))
        } else {
            Err(LineLengthFromIntError(value.get()))
        }
    }
}

impl std::fmt::Display for LineLengthFromIntError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(
            f,
            "The line length must be a value between 1 and {max}, not {value}.",
            max = LineLength::MAX,
            value = self.0
        )
    }
}

impl From<LineLength> for u16 {
    fn from(value: LineLength) -> Self {
        value.0.get()
    }
}

impl From<LineLength> for NonZeroU16 {
    fn from(value: LineLength) -> Self {
        value.0
    }
}

impl From<LineLength> for biome_formatter::LineWidth {
    fn from(value: LineLength) -> Self {
        // Unwrap: We assert that we match biome's `LineWidth` perfectly
        biome_formatter::LineWidth::try_from(value.value()).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use anyhow::Context;
    use anyhow::Result;

    use crate::settings::LineLength;

    #[derive(serde::Deserialize)]
    #[serde(deny_unknown_fields, rename_all = "kebab-case")]
    struct Options {
        line_length: Option<LineLength>,
    }

    #[test]
    fn deserialize_line_length() -> Result<()> {
        let options: Options = toml::from_str(
            r"
line-length = 50
",
        )?;

        assert_eq!(options.line_length, Some(LineLength::try_from(50).unwrap()));

        Ok(())
    }

    #[test]
    fn deserialize_oob_line_length() -> Result<()> {
        let result: std::result::Result<Options, toml::de::Error> = toml::from_str(
            r"
line-length = 400
",
        );
        let error = result.err().context("Expected OOB `LineLength` error")?;
        insta::assert_snapshot!(error);
        Ok(())
    }
}
