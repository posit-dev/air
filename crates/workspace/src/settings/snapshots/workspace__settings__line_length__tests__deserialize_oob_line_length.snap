---
source: crates/workspace/src/settings/line_length.rs
expression: error
---
TOML parse error at line 2, column 15
  |
2 | line-length = 400
  |               ^^^
The line length must be a value between 1 and 320, not 400.
